File: solutions/go/triangle/triangle.go
// Package triangle determines if a triangle is equilateral, isosceles, or scalene.
package triangle

// Type Kind is returned by function KindFromSides.
type Kind int

const (
	NaT = 0 // not a triangle
	Equ = 1 // equilateral
	Iso = 2 // isosceles
	Sca = 3 // scalene
)

// KindFromSides return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
func KindFromSides(a, b, c float64) (k Kind) {
	// all side lenghts must be positive
	if a <= 0 || b <= 0 || c <= 0 {
		return NaT
	}

	// side lengths cannot violate triangle inequality
	if a+b < c || b+c < a || a+c < b {
		return NaT
	}

	switch {
	case a == b && b == c:
		k = Equ
	case a == b, b == c, a == c:
		k = Iso
	default:
		k = Sca
	}
	return
}

File: solutions/r/triangle/triangle.R
type <- list(
    nat = c("not a triangle"),
    equ = c("equilateral", "isosceles"),
    iso = c("isosceles"),
    sca = c("scalene")
)

# triangle determines if a triangle is equilateral, isosceles, or scalene.
triangle <- function(x, y, z) {
    if (has_invalid_side(x, y, z) ||
        has_invalid_length(x, y, z)) {
        stop(type$nat)
    }

    if (x == y && y == z) {
        structure(c(x, y, z), class = type$equ)
    } else if (x == y || y == z || x == z) {
        structure(c(x, y, z), class = type$iso)
    } else {
        structure(c(x, y, z), class = type$sca)
    }
}

# all side lenghts must be positive
has_invalid_side <- function(x, y, z) {
    x <= 0 || y <= 0 || z <= 0
}

# side lengths cannot violate triangle inequality
has_invalid_length <- function(x, y, z) {
    x + y < z || y + z < x || x + z < y
}
